### PR TITLE
FormEditor SaveCanges Changes

### LINKED ISSUE #8347: [NBug] Access to the path 'C:\Users\lollo\.gitconfig' is denied.
## Error Details
```
System.UnauthorizedAccessException: Access to the path 'C:\Users\lollo\.gitconfig' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path, Boolean append, Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path, Boolean append, Encoding encoding, Int32 bufferSize, Boolean checkHost)
   at System.IO.File.InternalWriteAllText(String path, String contents, Encoding encoding, Boolean checkHost)
   at System.IO.File.WriteAllText(String path, String contents, Encoding encoding)
   at GitUI.CommandsDialogs.FormEditor.SaveChanges()
   at GitUI.CommandsDialogs.FormEditor.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at ResourceManager.GitExtensionsControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at ResourceManager.GitExtensionsControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.ContainerControl.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.ProcessCmdKey(Message& msg, Keys keyData)
   at System.Windows.Forms.Control.PreProcessMessage(Message& msg)
   at System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
   at System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)
```

## Additional information
File "~/.gitconfig" is hidden on my Windows install.

## Environment
- Git Extensions 3.4.2.9737
- Build 082cfd97d09c47346638b1873f5c8cf6ed139382
- Git 2.27.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by digitalcoyote**
Fixes #8347 

## Proposed changes

- Use the TextWriter Logic recommended on [StackOverflow](https://stackoverflow.com/questions/2246990/how-do-i-write-to-a-hidden-file/2247115#2247115)
   - Prevents issues with hidden files caused by File.WriteAllText
   - File.WriteAllBytes does not appear to have this issue based on reference source
- ~Move the general try/catch logic inside of SaveChanges()~
   - ~Added a paramter rethrowException (default: false) to allow custom handling such as on Form Close.~

## Test methodology (Not Tested yet, hence the **WIP** but the following will detail the testing plans)
-  Manual Testing Saving with no special circumstances
- Debugging to explicitly throw exceptions from inside SaveChanges to test rethrow on form close
- Manually Testing Saving Changes on ReadOnly Files, Hidden Files, and Files that are Locked ( I expect Readonly and Write Locked files to fail while hidden should succeed).

## Test environment(s) 

- GIT 2.28.0.windows.1
- Window 10 (Latest)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by digitalcoyote** [comment]
I pulled this together on lunch, so it's likely to be another 5+ hours before I can test it. That being said, there are some unit tests for this. I may add some to cover edge cases (such as hidden files) unless someone objects to it.

**Review by gerhardol — COMMENTED** [review]
SeemOK
There are a few other places files are written (config, gitignore, gitattributes etc) that maybe should be changed too, but I saw nothing urgent

**Comment by digitalcoyote** [comment]
I tend to prefer to have a light touch, but ill check around for anywhere that uses a file.writealltext in addition to any that don't have a try catch (as it appears there have been multiple bugs filed around this type of behavior).

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 1]
```diff
@@ -116,40 +116,52 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
-                }
-                else
+                if (!string.IsNullOrEmpty(_fileName))
                 {
-                    using (var bytes = new MemoryStream())
+                    if (fileViewer.FilePreamble is null || Module.FilesEncoding.GetPreamble().SequenceEqual(fileViewer.FilePreamble))
+                    {
+                        using FileStream fs = new FileStream(_fileName, FileMode.Open);
+                        using TextWriter tw = new StreamWriter(fs, Module.FilesEncoding, 1024, true);
+                        tw.Write(fileViewer.GetText());
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:136]:
```suggestion
                        using (TextWriter tw = new StreamWriter(fs, Module.FilesEncoding, bufferSize: 1024, leaveOpen: true))
                        {
                            tw.Write(fileViewer.GetText());
                        }
                        // after flushing, set the stream length to the current position in order to truncate leftover text
```

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 2]
```diff
@@ -116,40 +116,52 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
+                        fs.SetLength(fs.Position);
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:137]: This non-obvious workaround (including comments) should be factored out as a common helper function.
Where to put it, @RussKie?

**Review by mstv — CHANGES_REQUESTED** [review]

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 3]
```diff
@@ -116,40 +116,52 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
-            }
+            SaveChanges();
         }
 
-        private void SaveChanges()
+        /// <summary>
+        /// Saves Changes
+        /// </summary>
+        /// <param name="rethrowException">Allows Exception to be rethrown for custom handling if saving chang fails</param>
+        private void SaveChanges(bool rethrowException = false)
```
[the remark this answers is not in the document]
Comment (digitalcoyote) [reply GitUI/CommandsDialogs/FormEditor.cs:126]: I'm fine with waiting, I cannot imagine I'll be too busy to circle around to this when that discussion is finished.

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 2 above)
Comment (digitalcoyote) [reply GitUI/CommandsDialogs/FormEditor.cs:137]: I found another location that could benefit from it: ConfigFile's Save(). Other than that, the only other File.WriteAllText that is not in a test is for saving the commit message to a file (I assume these wouldn't normally be hidden).

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 4]
```diff
@@ -116,40 +116,60 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
+            catch (Exception ex)
+            {
+                if (rethrowException)
+                {
+                    throw;
                 }
 
-                // we've written the changes out to disk now, nothing to save.
-                HasChanges = false;
+                MessageBox.Show(this, _cannotSaveFile.Text + Environment.NewLine + ex.Message, Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormEditor.cs:172]:
```suggestion
                MessageBox.Show(this, $"{_cannotSaveFile.Text}{Environment.NewLine}{ex.Message}", Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
```

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 5]
```diff
@@ -116,40 +116,60 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
         {
-            if (!string.IsNullOrEmpty(_fileName))
+            try
             {
-                if (fileViewer.FilePreamble is null || Module.FilesEncoding.GetPreamble().SequenceEqual(fileViewer.FilePreamble))
-                {
-                    File.WriteAllText(_fileName, fileViewer.GetText(), Module.FilesEncoding);
-                }
-                else
+                if (!string.IsNullOrEmpty(_fileName))
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormEditor.cs:130]: Please invert to reduce nesting

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 6]
```diff
@@ -116,40 +116,60 @@ private void FormEditor_FormClosing(object sender, FormClosingEventArgs e)
...
-            try
-            {
-                SaveChanges();
-            }
-            catch (Exception ex)
-            {
-                MessageBox.Show(this, _cannotSaveFile.Text + Environment.NewLine + ex.Message, Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
-            }
+            SaveChanges();
         }
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormEditor.cs:127]: What was the reason to undo this try..catch, but retain it at the other callsite?
Doing this we increased the complexity of `SaveChanges` method, by handling error scenarios for `toolStripSaveButton_Click`, at the same time `FormEditor_FormClosing` doesn't benefit from these changes.

Wouldn't it better to retain try..catch at both respective call sites, and have `SaveChanges` only concentrate on saving?

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 5 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormEditor.cs:130]: This showed up as diff because of the indentation change, which will be reverted anyway (as we both requested).
So I'd say this refactoring is optional.

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 4 above)
Comment (digitalcoyote) [reply GitUI/CommandsDialogs/FormEditor.cs:172]: I copied this when I moved it. Depending on the comment below, would you like it changed if the try/catch move is reverted?I'd be adding it to the other location did not have a try/catch in that case as well.

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 6 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormEditor.cs:127]: > I originally did that because of [a comment that mentioned moving the try catch into SaveChanges()](https://github.com/gitextensions/gitextensions/issues/8347#issuecomment-663319449),

Ah, thank you for pointing. It's hard to keep track of all discussions (and I obviously don't remember things) :)
I think I had a cursory look at callsites but didn't dive in deeply enough to 

> but not all of the places it was called could be handled that way (the one that was left had logic to cancel closing the form).

In situations like this I generally try to discuss possible approaches to save efforts (although it is often useful to POC different approaches).

Since we can't generalise it with try/catch for all 3 callsites, it is probably better to add the missing try/catch. 

I'm not sure (at this stage) how  #8278 / #8358 will help, but if you happy to put this on ice, let's do it.

Thank you.

Comment (digitalcoyote) [reply GitUI/CommandsDialogs/FormEditor.cs:127]: > I'm not sure (at this stage) how #8278 / #8358 will help, but if you happy to put this on ice, let's do it.
Then I'd feel better about getting this in sooner. I doubt it's causing problems with many users, but if revisited specifically due to either of those I'd expect someone with a better understanding of those may be better at handling them.

I went ahead and reverted the try/catch changes, but wrapped the other call to SaveChanges to keep it from bubbling up (used the same MessageBox from the tooltip click's catch). I also switched them to StringInterpolation (as suggested)

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 7]
```diff
@@ -132,7 +132,17 @@ private void SaveChanges()
             {
                 if (fileViewer.FilePreamble is null || Module.FilesEncoding.GetPreamble().SequenceEqual(fileViewer.FilePreamble))
                 {
-                    File.WriteAllText(_fileName, fileViewer.GetText(), Module.FilesEncoding);
+                    using (var fs = new FileStream(_fileName, FileMode.Open))
+                    {
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:136]:
```suggestion
                    using var fs = new FileStream(_fileName, FileMode.Open);
```

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 8]
```diff
@@ -161,7 +171,15 @@ protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
...
                 case Keys.Control | Keys.S:
-                    SaveChanges();
+                    try
+                    {
+                        SaveChanges();
+                    }
+                    catch (Exception ex)
+                    {
+                        MessageBox.Show(this, $"{_cannotSaveFile.Text}{Environment.NewLine}{ex.Message}", Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
+                    }
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:181]: Do not copy & paste MessageBoxes. Factor out this to a new private function, e.g. `SaveChangesShowException`.
Do you have a better name, @RussKie?

File: GitUI/CommandsDialogs/FormEditor.cs (hunk 2 above)
Comment (mstv) [reply GitUI/CommandsDialogs/FormEditor.cs:144]: So factor out this function, please.
@RussKie, in which (new) class shall this be moved?

File: GitExtUtils/FileUtility.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
+﻿using System.IO;
+using System.Text;
+
+namespace GitUI
+{
+    public static class FileUtility
+    {
+        /// <summary>
+        ///     Writes all text to a file. Works around issues with Hidden Files encountered by File.WriteAllText
```
Comment (mstv) [inline GitExtUtils/FileUtility.cs:9]:
```suggestion
        /// Writes all text to a file. Works around issues with hidden files encountered by File.WriteAllText.
```

File: GitExtUtils/FileUtility.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        /// <param name="fileName">Destination file</param>
+        /// <param name="contents">Text to write as file contents</param>
+        public static void SafeWriteAllText(string fileName, string contents, Encoding encoding)
+        {
+            using (var fs = new FileStream(fileName, FileMode.Open))
+            {
+                using (TextWriter tw = new StreamWriter(fs, encoding, 1024,
+                    true))
```
Comment (mstv) [inline GitExtUtils/FileUtility.cs:18]:
```suggestion
                using (TextWriter tw = new StreamWriter(fs, encoding, bufferSize: 4096, leaveOpen: true))
```

File: GitUI/CommandsDialogs/FormEditor.cs [hunk 11]
```diff
@@ -122,17 +122,22 @@ private void toolStripSaveButton_Click(object sender, EventArgs e)
...
             catch (Exception ex)
             {
-                MessageBox.Show(this, _cannotSaveFile.Text + Environment.NewLine + ex.Message, Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
+                SaveChangesShowException(ex);
             }
         }
 
+        private void SaveChangesShowException(Exception ex)
+        {
+            MessageBox.Show(this, $"{_cannotSaveFile.Text}{Environment.NewLine}{ex.Message}", Strings.Error, MessageBoxButtons.OK, MessageBoxIcon.Error);
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormEditor.cs:131]:
```suggestion
            try
            {
                SaveChanges();
            }
            catch (ex)
            {
                MessageBoxes.ShowError(this, $"{_cannotSaveFile.Text}{Environment.NewLine}{ex.Message}");
            }
```

**Review by mstv — COMMENTED** [review]
Almost there

**Review by RussKie — COMMENTED** [review]
I think we can take it 👍

**Review by mstv — APPROVED** [review]

**Comment by mstv** [comment]
@msftbot merge in 1 day

**Comment by RussKie** [comment]
Thank you
